pkg/p2p/message_router.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package p2p
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/ticdc/pkg/security"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+// MessageRouter is used to maintain clients to all the peers in the cluster
+// that the local node needs to communicate with.
+type MessageRouter interface {
+	// AddPeer should be invoked when a new peer is discovered.
+	AddPeer(id NodeID, addr string)
+	// RemovePeer should be invoked when a peer is determined to
+	// be permanently unavailable.
+	RemovePeer(id NodeID)
+	// GetClient returns a MessageClient for `target`. It returns
+	// nil if the target peer does not exist. The returned client
+	// is canceled if RemovePeer is called on `target`.
+	GetClient(target NodeID) *MessageClient
+	// Close cancels all clients maintained internally.
+	Close()
+	// Wait waits for all clients to exit.
+	Wait()
+	// Err returns a channel to receive errors from.
+	Err() <-chan error
+}
+
+type messageRouterImpl struct {
+	mu         sync.RWMutex
+	addressMap map[NodeID]string
+	clients    map[NodeID]clientWrapper
+
+	wg       sync.WaitGroup
+	isClosed atomic.Bool
+	errCh    chan error
+
+	// read only field
+	credentials  *security.Credential
+	selfID       NodeID
+	clientConfig *MessageClientConfig
+}
+
+// NewMessageRouter creates a new MessageRouter
+func NewMessageRouter(selfID NodeID, credentials *security.Credential, clientConfig *MessageClientConfig) MessageRouter {
+	return &messageRouterImpl{
+		addressMap:   make(map[NodeID]string),
+		clients:      make(map[NodeID]clientWrapper),
+		errCh:        make(chan error, 1), // one error at most
+		credentials:  credentials,
+		selfID:       selfID,
+		clientConfig: clientConfig,
+	}
+}
+
+type clientWrapper struct {
+	*MessageClient
+	cancelFn context.CancelFunc
+}
+
+// AddPeer implements MessageRouter.
+func (m *messageRouterImpl) AddPeer(id NodeID, addr string) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.addressMap[id] = addr
+}
+
+// RemovePeer implements MessageRouter.
+func (m *messageRouterImpl) RemovePeer(id NodeID) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	delete(m.addressMap, id)
+	if clientWrapper, ok := m.clients[id]; ok {
+		clientWrapper.cancelFn()
+	}
```

Comment:
Does it need to remove client(id) from `m.clients`?

Reply:
The client is removed from `m.clients` only after it is canceled, to prevent two clients to the same target from co-exist (although this is tolerated to same degree).